src/docfx/restore/Restore.cs

```diff
@@ -41,181 +37,41 @@ internal static class Restore
-
             var result = new RestoreLock();
-            foreach (var (href, workTreeHead) in workTreeMappings)
-            {
-                result.Git[href] = workTreeHead;
-            }
-
-            // todo: restore other items, @renze
-            return result;
-
-            async Task<List<(string href, string head)>> RestoreDependentRepo(string restoreDir, List<string> hrefs)
-            {
-                var workTreeHeads = await GetWorkTrees(docsetPath, restoreDir, hrefs);
-
-                foreach (var (_, workTreeHead) in workTreeHeads)
-                {
-                    var childDir = GetRestoreWorkTreeDir(restoreDir, workTreeHead);
-                    await restoreChild(childDir);
-                }
-
-                return workTreeHeads;
-            }
-        }
-
-        // Restore dependent repo to local and create work tree
-        private static async Task<List<(string href, string head)>> GetWorkTrees(string docsetPath, string restoreDir, List<string> hrefs)
-        {
-            var restorePath = PathUtility.NormalizeFolder(Path.Combine(restoreDir, ".git"));
-            var (url, _) = GetGitRemoteInfo(hrefs.First());
-            var workTreeHeads = new ConcurrentBag<(string href, string head)>();
-
-            await ProcessUtility.CreateFileMutex(
-                PathUtility.NormalizeFile(Path.GetRelativePath(AppData.RestoreDir, restorePath)),
-                async () =>
-                {
-                    await FetchOrCloneRepo();
-
-                    var (existingWorkTrees, newWorkTrees) = await AddWorkTrees();
-
-                    if (NeedCleanupWorkTrees(existingWorkTrees.Count))
-                    {
-                        using (Progress.Start($"Cleanup `{restoreDir}` work trees"))
-                        {
-                            await CleanupWorkTrees(existingWorkTrees, newWorkTrees);
-                        }
-                    }
-                });
 
-            async Task FetchOrCloneRepo()
+            // restore git dependnecy repositories
+            var workTreeHeadMappings = await RestoreGit.Restore(docsetPath, config, restoreChild);
+            foreach (var (href, workTreeHead) in workTreeHeadMappings)
             {
-                if (GitUtility.IsRepo(restoreDir))
-                {
-                    // already exists, just pull the new updates from remote
-                    await GitUtility.Fetch(restorePath);
-                }
-                else
-                {
-                    // doesn't exist yet, clone this repo to a specified branch
-                    await GitUtility.Clone(restoreDir, url, restorePath, null, true);
-                }
+                result.Git[href] = workTreeHead;
             }
 
-            async Task<(List<string> existingWorkTrees, List<string> newWorkTrees)> AddWorkTrees()
+            // restore github user
+            if (!string.IsNullOrEmpty(config.Contribution.UserProfileCacheUrl))
```

Comment:
>IsNullOrEmpty [](start = 24, length = 13)

Add a method `IEnumerable<string> Config.GetRestoreUrls()`

Reply:
OK

---
In reply to: [202252344](https://github.com/dotnet/docfx/pull/3065#discussion_r202252344) [](ancestors = 202252344)

Reply:
disagree, the config has nothing to do with restore, we should manage all restore items inside of restore

---
In reply to: [202591479](https://github.com/dotnet/docfx/pull/3065#discussion_r202591479) [](ancestors = 202591479,202302121,202252344)